pkg/virt-handler/vm.go

```diff
@@ -1422,12 +1423,52 @@ func (d *VirtualMachineController) calculateLiveMigrationCondition(vmi *v1.Virtu
 		return newNonMigratableCondition("VMI uses virtiofs", v1.VirtualMachineInstanceReasonVirtIOFSNotMigratable), isBlockMigration
 	}
 
+	containsPCIHostDev, err := vmiContainsPCIHostDevice(vmi)
+	// should never happen; checking error in case the underline implementation will be changed
+	if err != nil {
+		return newNonMigratableCondition("can't get host devices", v1.VirtualMachineInstanceReasonHostDeviceNotMigratable), isBlockMigration
+	}
+
+	if containsPCIHostDev {
+		return newNonMigratableCondition("VMI uses a PCI host devices", v1.VirtualMachineInstanceReasonHostDeviceNotMigratable), isBlockMigration
+	}
+
 	return &v1.VirtualMachineInstanceCondition{
 		Type:   v1.VirtualMachineInstanceIsMigratable,
 		Status: k8sv1.ConditionTrue,
 	}, isBlockMigration
 }
 
+// Note: genhostdev.CreateHostDevices uses NewBestEffortAddressPool, that its Pop() function does not return error; the
+// meaning is that this function won't return error unless the implementation of the genhostdev.CreateHostDevices function
+// will be changed.
+func vmiContainsPCIHostDevice(vmi *v1.VirtualMachineInstance) (bool, error) {
+
+	if numDevs := len(vmi.Spec.Domain.Devices.HostDevices) + len(vmi.Spec.Domain.Devices.GPUs); numDevs > 0 {
+		devices := make([]v1.HostDevice, numDevs)
+		copy(devices, vmi.Spec.Domain.Devices.HostDevices)
+		// translate GPU devices to Host Devices
+		for i, gpu := range vmi.Spec.Domain.Devices.GPUs {
+			devices[i+len(vmi.Spec.Domain.Devices.HostDevices)] = v1.HostDevice{
+				Name:       gpu.Name,
+				DeviceName: gpu.DeviceName,
+			}
```

Comment:
why aren't you using `append` here?

Reply:
Because append cost in case of reallocation. I never use append for slices with a known length.